### PR TITLE
[REBASE][FF-A] Ported MS Services to UEFI Style

### PR DESCRIPTION
**Pull request by Raymond-MS**
## Description

All services from Trusted Services have been ported to UEFI style and are running in a secure partition generated our UEFI style as well. The services have been ported as libraries that secure partitions can include if they would like to utilize them. Added an ArmArchTimer library for the secure partition to support the TPM library for the TPM service. SEL0 SP's do not have access to timer registers so we needed a work-around.

For details on how to complete these options and their meaning refer to [CONTRIBUTING.md](https://github.com/microsoft/mu/blob/HEAD/CONTRIBUTING.md).

- [x] Impacts functionality?
- [x] Impacts security?
- [ ] Breaking change?
- [ ] Includes tests?
- [ ] Includes documentation?
- [ ] Backport to release branch?

## How This Was Tested

Verified service functionality through the FfaPartitionTestApp with the TPM service being tested through building with the TPM defines set in the build_conf file. Secure partition is being dispatched and all services are being exercised. 

## Integration Instructions

N/A

### CODE REVIEW

File: ArmPkg/Library/TpmServiceLib/TpmServiceLib.c [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+VOID
+TpmServiceDeInit (
+  VOID
+  )
+{
+  /* Nothing to DeInit */
+}
+
+VOID
+TpmServiceHandle (
```
Comment (kuqin12) [inline ArmPkg/Library/TpmServiceLib/TpmServiceLib.c:435]: I think these libraries are well architected to be in separate commits.

Let's break this PR into 3 commits (maybe 4 if we stick to the timer lib you have here), one for each service.

File: ArmPkg/Library/TpmServiceLib/TpmServiceLib.c [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+}
+
+STATIC
+TpmStatus
+FinishHandler (
+  DIRECT_MSG_ARGS_EX *Request, 
+  DIRECT_MSG_ARGS_EX *Response
+  )
+{
+  (void)Request;
```
Comment (kuqin12) [inline ArmPkg/Library/TpmServiceLib/TpmServiceLib.c]: What is this line for?

File: ArmPkg/Library/TpmServiceLib/TpmServiceLib.c [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+}
+
+STATIC
+TpmStatus
+UnregisterHandler (
+  DIRECT_MSG_ARGS_EX *Request, 
+  DIRECT_MSG_ARGS_EX *Response
+  )
+{
+  (void)Request;
```
Comment (kuqin12) [inline ArmPkg/Library/TpmServiceLib/TpmServiceLib.c]: Same here?

File: ArmPkg/Library/TpmServiceLib/TpmServiceLib.c [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+}
+
+STATIC
+TpmStatus
+RegisterHandler (
+  DIRECT_MSG_ARGS_EX *Request, 
+  DIRECT_MSG_ARGS_EX *Response
+  )
+{
+  (void)Request;
```
Comment (kuqin12) [inline ArmPkg/Library/TpmServiceLib/TpmServiceLib.c]: ditto

File: ArmPkg/Library/TpmServiceLib/TpmServiceLib.c (hunk 2 above)
Comment (kuqin12) [reply ArmPkg/Library/TpmServiceLib/TpmServiceLib.c]: Also do you need these handler related functions in this context?

File: ArmPkg/Library/TpmServiceLib/TpmServiceLib.c [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+/* TPM Service Defines */
+#define TPM_MAJOR_VER              (1)
+#define TPM_MINOR_VER              (0)
+
+#define TPM_START_PROCESS_CMD      (0)
+#define TPM_START_PROCESS_LOC_REQ  (1)
+
+#define TPM_LOC_STATE_MASK         (0x9F)
+#define TPM_LOC_STS_MASK           (0x01)
+#define TPM_CTRL_START_MASK        (0x01)
```
Comment (kuqin12) [inline ArmPkg/Library/TpmServiceLib/TpmServiceLib.c]: nit: do you think it makes sense to have these live in the lib header?

Comment (Raymond-MS) [reply ArmPkg/Library/TpmServiceLib/TpmServiceLib.c]: No because they aren't needed outside of the .c file. These are only needed by the cleanup function which makes sure the CRB is in a good state.

File: ArmPkg/Library/TpmServiceLib/TpmServiceLib.c (hunk 4 above)
Comment (Raymond-MS) [reply ArmPkg/Library/TpmServiceLib/TpmServiceLib.c]: Removed. This was due to an unused parameter in TS and was copied here.

File: ArmPkg/Library/TpmServiceLib/TpmServiceLib.c (hunk 3 above)
Comment (Raymond-MS) [reply ArmPkg/Library/TpmServiceLib/TpmServiceLib.c]: Removed

File: ArmPkg/Library/TpmServiceLib/TpmServiceLib.c (hunk 2 above)
Comment (Raymond-MS) [reply ArmPkg/Library/TpmServiceLib/TpmServiceLib.c]: Removed. And yes, I think it is worth keeping them stubbed out in case we need to implement them in the future. This doesn't hurt and keeps things clear rather than just returning an error code in the handler.

File: ArmPkg/Library/TpmServiceLib/TpmServiceLib.c [hunk 6]
```diff
@@ -0,0 +1,471 @@
...
+  )
+{
+  /* TODO: If we end up supporting more than Locality0, this will need to init all localities supported */
+  PTP_CRB_REGISTERS_PTR  InternalTpmCrb;
+
+  InternalTpmCrb = (PTP_CRB_REGISTERS_PTR)(UINTN)PcdGet64 (PcdTpmInternalBaseAddress);
+  DEBUG ((DEBUG_INFO, "PcdTpmInternalBaseAddress: %lx\n", PcdGet64 (PcdTpmInternalBaseAddress)));
+  SetMem ((void *)InternalTpmCrb, sizeof (PTP_CRB_REGISTERS), 0x00);
+  InternalTpmCrb->LocalityState = 0x01;   // Default Value - tpmEstablished
+  InternalTpmCrb->InterfaceId   = 0x4011; // CRB Interface Selected, Only Locality0 Supported
```
Comment (kuqin12) [inline ArmPkg/Library/TpmServiceLib/TpmServiceLib.c]: Consider using #define?

File: ArmPkg/Include/Library/TpmServiceLib.h [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+#define TPM_STATUS_INV_CRB_CTRL_DATA  ((TpmStatus)0x8e000006)
+#define TPM_STATUS_ALREADY            ((TpmStatus)0x8e000009)
+#define TPM_STATUS_DENIED             ((TpmStatus)0x8e00000A)
+#define TPM_STATUS_NOMEM              ((TpmStatus)0x8e00000B)
+
+/* Opcodes
+ * TPM Opcodes spec defined in: https://developer.arm.com/documentation/den0138/0100/?lang=en */
+#define TPM_OPCODE_BASE                   (0x0F000000)
+#define TPM_OPCODE_GET_INTERFACE_VERSION  (TPM_OPCODE_BASE + 0x01)
+#define TPM_OPCODE_GET_FEATURE_VERSION    (TPM_OPCODE_BASE + 0x101)
```
Comment (kuqin12) [inline ArmPkg/Include/Library/TpmServiceLib.h]: This is already defined in the `Guid/Tpm2ServiceFfa.h`. Probably should reuse those?

File: ArmPkg/Include/Library/NotificationServiceLib.h [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+
+**/
+
+#ifndef NOTIFICATION_SERVICE_LIB_H_
+#define NOTIFICATION_SERVICE_LIB_H_
+
+#include <Base.h>
+#include <IndustryStandard/ArmFfaPartInfo.h>
+#include <Library/ArmSvcLib.h>
+#include <Library/ArmFfaLibEx.h>
```
Comment (kuqin12) [inline ArmPkg/Include/Library/NotificationServiceLib.h:15]: nit: we generally do not recommend carrying header files in the header files. But we can worry about it later.

File: ArmPkg/Library/NotificationServiceLib/NotificationServiceLib.c [hunk 9]
```diff
@@ -0,0 +1,518 @@
...
+} NotifBits;
+
+typedef struct  {
+  UINT8        ServiceUuid[16];
+  NotifBits    ServiceBits[NOTIFICATION_MAX_IDS];
+  BOOLEAN      InUse;
+} NotifService;
+
+/* Notification Service Variables */
+static BOOLEAN       IdsAcquired                                     = FALSE;
```
Comment (kuqin12) [inline ArmPkg/Library/NotificationServiceLib/NotificationServiceLib.c]: Use `STATIC`?

File: ArmPkg/Library/ArmArchTimerLibEx/ArmArchTimerLibEx.inf [hunk 10]
```diff
@@ -0,0 +1,32 @@
+#/** @file
+#
+#  Copyright (c) 2011 - 2014, ARM Limited. All rights reserved.<BR>
+#  SPDX-License-Identifier: BSD-2-Clause-Patent
+#
+#**/
+
```
Comment (kuqin12) [inline ArmPkg/Library/ArmArchTimerLibEx/ArmArchTimerLibEx.inf:7]: Probably want to add an override tag for this:https://github.com/microsoft/mu_basecore/blob/dev/202405/BaseTools/Plugin/OverrideValidation/ReadMe.md

**Review by kuqin12 — APPROVED** [review]
